### PR TITLE
Opmphm integration1

### PR DESCRIPTION
**Pull request by KurtMi**
# Purpose

The new opmphm

# Checklist

- [x] commit messages are fine (with references to issues)
- [x] I ran all tests and everything went fine
- [x] I added unit tests
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions

### CODE REVIEW

**Comment by KurtMi** [comment]
jenkins build gcc-configure-debian-optimizations please

**Comment by KurtMi** [comment]
jenkins build gcc-configure-debian-optimizations please

**Comment by KurtMi** [comment]
jenkins build gcc-configure-debian-optimizations please

**Comment by KurtMi** [comment]
@markus2330 please review this PR.

File: doc/dev/data-structures.md [hunk 1]
```diff
@@ -369,34 +369,29 @@ All structs are defined in [opmphm.h](/src/include/kdbopmphm.h).
...
 Provide a good seed in `OpmphmInit->initSeed`, needed in the next step.
-Call the `opmphmNewOrder ()` and use the default order or your own order.
-Min and Max order also need to be set respectively.
 
-The `opmphmInit ()`, transforms the orders in a internal representation, this step
-is irreversible.
+### Mapping
 
-### The Mapping
+The `opmphmMapping ()` uses your seed (the `OpmphmInit->seed` will be changed) and tries to map your
```
Comment (markus2330) [inline doc/dev/data-structures.md]: I think we can drop the (), it somehow suggests that there would be no parameter. Maybe:
```
The function `opmphmMapping` 
```
is more clear?

File: doc/dev/data-structures.md [hunk 2]
```diff
@@ -369,34 +369,29 @@ All structs are defined in [opmphm.h](/src/include/kdbopmphm.h).
...
+elements to edges in a acyclic r-partite hypergraph, this mapping might not succeed, on cycles just call it again.
 
-The `opmphmMapping ()` uses your seed (the OpmphmInit->seed will be changed) and tries to map your
-elements to a fixed space, this mapping might not succeed, on duplicate just call it again.
+### Assignment
 
-### The Build
+The `opmphmAssignment ()` function assigns either your order or a default order. The default order is the order
```
Comment (markus2330) [inline doc/dev/data-structures.md]: You removed opmphmNewOrder from the docu here?

File: scripts/benchmark_opmphm [hunk 3]
```diff
@@ -13,7 +13,7 @@
 # Note: feed exact amount of needed seeds otherwise you get a SIGPIPE
 #
 
-maxOPMPHMTUPLE=10
+maxOPMPHMR_PARTITE=6
```
Comment (markus2330) [inline scripts/benchmark_opmphm]: What about all upper `MAX_OPMPHMR_PARTITE`?

File: scripts/draw-opmphm [hunk 4]
```diff
@@ -1,14 +0,0 @@
-#!/bin/sh
```
Comment (markus2330) [inline scripts/draw-opmphm]: Is it still possible to draw it?

File: src/include/kdbopmphm.h [hunk 5]
```diff
@@ -8,98 +8,93 @@
...
 /**
- * The Order Preserving Minimal Perfect Hash Map (OPMPHM) first maps each key name to a k-tuple, k is stored in OPMPHMTUPLE.
- * Each element of the k-tuples is in the range 0 to width. The width (abbreviated with w) determines how many keys can fit in the OPMPHM.
- * w^OPMPHMTUPLE is the maximum number of keys that fit in. w is set dynamically to the minimum value that:
- *
- * w^OPMPHMTUPLE >= n * n / opmphmRatio
- *
- * opmphmRatio should never be less or equal to 0.
- *
+ * The Order Preserving Minimal Perfect Hash Map (opmphm) maps each element to an edge in a r-partite hypergraph.
```
Comment (markus2330) [inline src/include/kdbopmphm.h]: How do you pronounce r-partite?
I would consider writing "an r-partite".

File: src/include/kdbopmphm.h [hunk 6]
```diff
@@ -8,98 +8,93 @@
...
  */
 typedef struct
 {
-	size_t h[OPMPHMTUPLE]; /*!< the k-tuple filled by the OPMPHM with the hash function values */
-	union {
-		size_t p;	      /*!< external usage*/
-		size_t t[OPMPHMTUPLE]; /*!< internal usage */
-	} index;		       /*!< desired hash map return value */
-} OpmphmOrder;
+	size_t * g;					  /*!< array containing the final opmphm */
```
Comment (markus2330) [inline src/include/kdbopmphm.h]: `g` is really short.

File: src/include/kdbopmphm.h [hunk 7]
```diff
@@ -8,98 +8,93 @@
...
+	size_t size;					  /*!< size of g in bytes */
+	int32_t opmphmHashFunctionSeeds[OPMPHMR_PARTITE]; /*!< the seed for the hash function calls */
+	size_t p;					  /*!< the number of vertices in one part of the r-partite hypergraph */
```
Comment (markus2330) [inline src/include/kdbopmphm.h]: ditto.

File: src/libs/elektra/opmphm.c [hunk 8]
```diff
@@ -15,220 +15,354 @@
...
+ * @retval 0 on success
+ * @retval -1 on memory error
+ */
+int opmphmAssignment (Opmphm * opmphm, OpmphmGraph * graph, size_t n, int defaultOrder)
+{
+	ELEKTRA_ASSERT (opmphm != NULL, "passed opmphm is a Null Pointer");
+	ELEKTRA_ASSERT (graph != NULL, "passed graph is a Null Pointer");
+	ELEKTRA_ASSERT (graph->removeIndex == n, "graph contains a cycle");
+	ELEKTRA_ASSERT (n > 0, "n is 0");
+	// malloc g
```
Comment (markus2330) [inline src/libs/elektra/opmphm.c]: This is quite obvious but other non-commented parts are not.

File: src/libs/elektra/opmphm.c [hunk 9]
```diff
@@ -15,220 +15,354 @@
...
-			offset[i] = nextOffset[i - 1];
-		}
-		offset[0] = 0;
-		for (size_t i = 1; i < w; ++i)
-		{
-			offset[i] += offset[i - 1];
-		}
-		// collection
-		for (size_t i = 0; i < n; ++i)
+		ELEKTRA_ASSERT (notAssignedCount != 0, "no not Assigned vertex");
```
Comment (markus2330) [inline src/libs/elektra/opmphm.c]: lower case in assigned?

File: tests/ctest/test_opmphm.c [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+static const char * test_opmphm_getString (void * data ELEKTRA_UNUSED)
+{
+	return "Test Key Name";
+}
+
+const size_t maxP = 1000;
+
+static void test_cyclic (void)
+{
+	{
```
Comment (markus2330) [inline tests/ctest/test_opmphm.c]: What putting them in extra functions?

`test_cyclicNameForSpecificTest`

File: tests/ctest/test_opmphm.c [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+{
+	printf ("OPMPHM      TESTS\n");
+	printf ("==================\n\n");
+
+	init (argc, argv);
+
+	test_cyclic ();
+	test_acyclic ();
+
+	printf ("\ntest_key RESULTS: %d test(s) done. %d error(s).\n", nbTest, nbError);
```
Comment (markus2330) [inline tests/ctest/test_opmphm.c]: key -> opmphm

File: scripts/draw-opmphm (hunk 4 above)
Comment (KurtMi) [reply scripts/draw-opmphm]: It seems that dot files do not support hypergraphs.

File: doc/dev/data-structures.md (hunk 2 above)
Comment (KurtMi) [reply doc/dev/data-structures.md]: The `OpmphmOrder` struct is not needed an more. Some fields went to `OpmphmEdge`.

**Comment by KurtMi** [comment]
The parts of the data-structures and benchmarks are not 100% finished.

**Comment by KurtMi** [comment]
jenkins build gcc-configure-debian-optimizations please

**Comment by KurtMi** [comment]
jenkins build gcc-configure-debian-optimizations please

**Comment by KurtMi** [comment]
The data-structures.md and the OPMPHM implementation should now be consistent.

**Comment by markus2330** [comment]
Thank you, great! I hope I soon get something to read about your work!

**Pull request merged**